libsolidity/formal/CHC.cpp

```diff
@@ -884,9 +893,15 @@ void CHC::defineContractInitializer(ContractDefinition const& _contract)
 	addRule(smtutil::Expression::implies(initialConstraints(_contract), implicitFact), implicitFact.name);
 	setCurrentBlock(implicitConstructorPredicate);
 
+	solAssert(!m_errorDest, "");
+	m_errorDest = m_contractInitializers.at(&_contract);
 	for (auto var: _contract.stateVariables())
-		m_context.variable(*var)->increaseIndex();
-	initializeStateVariables(_contract);
+		if (var->value())
+		{
+			var->value()->accept(*this);
+			assignment(*var, *var->value());
+		}
```

Comment:
@mijovic This is the fix for the variable initialization order. All state variables are already initialized as 0 before the whole inheritance deployment starts, which is correct. The wrong part here is that `SMTEncoder::initializeStateVariables` (removed line) first zero initializes then assigns values if available at state variable declaration. This is wrong here, since the zero init could overwrite other previous assignments which are valid. The correct behavior here is to only assign the declaration values, if available. In that case it is correct to overwrite previous assignments.